tagmanager/src/tm_source_file.c

```diff
@@ -223,20 +245,34 @@ static void tm_source_file_destroy(TMSourceFile *source_file)
 	source_file->tags_array = NULL;
 }
 
-/** Frees a TMSourceFile structure, including all contents. Before calling this
- function the TMSourceFile has to be removed from the TMWorkspace. 
- @param source_file The source file to free.
+/** Decrements the reference count of @a source_file
+ *
+ * If the reference count drops to 0, then @a source_file is freed, including all contents.
+ * Make sure the @a source_file is already removed from any TMWorkSpace before the
+ * this happens.
+ * @param source_file The source file to free.
+ * @see tm_workspace_remove_source_file()
 */
 GEANY_API_SYMBOL
 void tm_source_file_free(TMSourceFile *source_file)
 {
-	if (NULL != source_file)
+	TMSourceFilePriv *priv = (TMSourceFilePriv *) source_file;
+
+	if (NULL != priv && g_atomic_int_dec_and_test(&priv->refcount))
 	{
 		tm_source_file_destroy(source_file);
-		g_free(source_file);
+		SOURCE_FILE_FREE(priv);
 	}
 }
 
+/** Gets the GBoxed-derived GType for TMSourceFile
+ *
+ * @return TMSourceFile type . */
+GEANY_API_SYMBOL
+GType tm_source_file_get_type(void);
```

Comment:
```gcc
../../../tagmanager/src/tm_source_file.c:272:7: warning: redundant redeclaration of 'tm_source_file_get_type' [-Wredundant-decls]
 GType tm_source_file_get_type(void);
       ^
In file included from ../../../tagmanager/src/tm_source_file.c:35:0:
../../../tagmanager/src/tm_source_file.h:48:7: note: previous declaration of 'tm_source_file_get_type' was here
 GType tm_source_file_get_type(void);
       ^
```
but I guess we don't care.  Alternatively we could just rely on the fact `G_DEFINE_BOXED_TYPE` only expands to a function definition so attributes could come right before, but this is likely to be fragile as it's not documented explicitly.